sycl/include/sycl/ext/oneapi/properties/properties.hpp

```diff
@@ -9,265 +9,194 @@
+// create map entry for property value
+// base-case for runtime property: entry is just key (=value)
+template <class PropT> struct make_entry {
+  using type = type_list<PropT>;
 };
-template <typename T, typename... Ts>
-struct RuntimePropertyStorage<std::tuple<T, Ts...>>
-    : std::conditional_t<IsRuntimeProperty<T>::value,
-                         PrependTuple<T, typename RuntimePropertyStorage<
-                                             std::tuple<Ts...>>::type>,
-                         RuntimePropertyStorage<std::tuple<Ts...>>> {};
-
-// Helper class to extract a subset of elements from a tuple.
-// NOTES: This assumes no duplicate properties and that all properties in the
-//        struct template argument appear in the tuple passed to Extract.
-template <typename PropertiesT> struct ExtractProperties {};
-template <typename... PropertiesTs>
-struct ExtractProperties<std::tuple<PropertiesTs...>> {
-  template <typename... PropertyValueTs>
-  using ExtractedPropertiesT = std::tuple<>;
-
-  template <typename... PropertyValueTs>
-  static constexpr ExtractedPropertiesT<PropertyValueTs...>
-  Extract(std::tuple<PropertyValueTs...>) {
-    return {};
-  }
-};
-template <typename PropertyT, typename... PropertiesTs>
-struct ExtractProperties<std::tuple<PropertyT, PropertiesTs...>> {
-  template <typename... PropertyValueTs>
-  using NextExtractedPropertiesT =
-      typename ExtractProperties<std::tuple<PropertiesTs...>>::
-          template ExtractedPropertiesT<PropertyValueTs...>;
-  template <typename... PropertyValueTs>
-  using ExtractedPropertiesT =
-      typename PrependTuple<PropertyT,
-                            NextExtractedPropertiesT<PropertyValueTs...>>::type;
-
-  template <typename... PropertyValueTs>
-  static constexpr ExtractedPropertiesT<PropertyValueTs...>
-  Extract(std::tuple<PropertyValueTs...> PropertyValues) {
-    PropertyT ThisExtractedProperty = std::get<PropertyT>(PropertyValues);
-    NextExtractedPropertiesT<PropertyValueTs...> NextExtractedProperties =
-        ExtractProperties<std::tuple<PropertiesTs...>>::template Extract<
-            PropertyValueTs...>(PropertyValues);
-    return std::tuple_cat(std::tuple<PropertyT>{ThisExtractedProperty},
-                          NextExtractedProperties);
-  }
+// specialization for compile-time property
+// property value is already a list with key as first type
+template <class... T> struct make_entry<property_value<T...>> {
+  using type = property_value<T...>;
 };
-
+template <class PropT> using make_entry_t = typename make_entry<PropT>::type;
+// create property map from properties type
+template <class PropertiesT>
+using property_map =
+    mp11::mp_transform<make_entry_t, mp11::mp_front<PropertiesT>>;
+template <class T, class = void>
+struct is_empty_or_incomplete : std::true_type {};
+template <class T>
+struct is_empty_or_incomplete<
+    T, std::enable_if_t<(sizeof(T) > 0) && !std::is_empty_v<T>>>
+    : std::false_type {};
```

Comment:
Did we do the same thing before? I'm worried about "incomplete" part and issues it may cause if the behavior is different when we ask for the same thing both before and after something become "complete" (and that would also be ODR-violation, I guess).

Reply:
No. And it's unrelated to the storage change. It is related to https://github.com/intel/llvm/pull/13669 which suggests that we allow to define properties as:
```
inline constexpr property_value<struct bar_key> bar;
```
If we don't allow incomplete we need to require it to be:
```
struct bar_key {};
inline constexpr property_value<bar_key> bar;
```
It has no other advantage besides being less verbose. So if we think this might be a problem we don't need to do this.
Note that it wouldn't be a concern that it would become complete. Because such a key would always just be forward declared.